platforms/documentation/docs/src/docs/userguide/authoring-builds/build_environment.adoc

```diff
@@ -131,54 +129,82 @@ spring-boot.version = 2.2.1.RELEASE
 [[sec:project_properties]]
 == Project properties
 
-Project properties are specific to your Gradle project, they are variables or configuration blocks defined in your build script or in the `project` object.
+Project properties are specific to your Gradle project, they are variables or configuration blocks defined in your build script or in the `Project` object.
+Project properties can be retrieved:
 
-For example, the project property `dryRun`, in the build file below, enables a specific test mode:
+- Lazily: `providers.gradleProperty()`
+- Eagerly: `project.ext()`, `project.findProperty()`, `project.property()`, `project.('<propertyName>')`
```

Comment:
I think the docs should be exhaustive because users may see these in the wild and come to the docs looking for an explanation, but this really highlights one of the main complaints which is that there's just too many ways to do things.

Do we have a preferred way to reference properties? My guess would be lazily via `providers.gradleProperty()`. If you really need a value eagerly you can call `providers.gradleProperty(...).get()`.

It would be nice to just say that the lazy method is the recommended way but you might see these other ways in older projects.